박규동/4월 4주차/Main_1938.java

```diff
@@ -0,0 +1,140 @@
+import java.util.*;
+import java.io.*;
+
+public class Main_1938 {
+
+	static int N,answer;
+	static char[][] board;
+	static boolean[][][] visit;
+	static int[] dy = { 0, 0, 1, -1,1,1,-1,-1 };
+	static int[] dx = { 1, -1, 0, 0,-1,1,-1,1 };
+	static int[] lcr = { -1, 0, 1 };
+	static Namu[] init,finish;
+	
+	static class Namu {
+		int y;
+		int x;
+		int state;
+		int dist;
+
+		public Namu(int y, int x, int state) {
+			this.y = y;
+			this.x = x;
+			this.state = state;
+		}
+		
+		public Namu(int y, int x, int state,int dist) {
+			this.y = y;
+			this.x = x;
+			this.state = state;
+			this.dist = dist;
+		}
+		
+	}
+
+	private static void bfs(Namu namu) {
+		visit[namu.y][namu.x][namu.state] = true;
+		int sy = namu.y;
+		int sx = namu.x;
+		int sstate = namu.state;
+		
+		Queue<Namu> q = new LinkedList<>();
+		q.add(new Namu(sy,sx,sstate,0));
+		while (!q.isEmpty()) {
+			int cy = q.peek().y;
+			int cx = q.peek().x;
+			int cstate = q.peek().state;
+			int cdist = q.peek().dist;
+			q.poll();
+			if(cy==finish[1].y&&cx==finish[1].x&&cstate==finish[1].state) {
+				answer = cdist;
+				return;
+			}
+			
+			
+			for (int i = 0; i < 4; i++) {
+				int ny = cy + dy[i];
+				int nx = cx + dx[i];
+				if(!ischeck(ny,nx,cstate)) continue;
+				visit[ny][nx][cstate]=true;
+				q.add(new Namu(ny, nx, cstate,cdist+1));
+			}
+			int nstate = 0;
+			if(cstate==0) nstate=1;
+			
+			if(ischeckRotate(cy,cx,nstate)) {
+				visit[cy][cx][nstate]=true;
+				q.add(new Namu(cy, cx, nstate,cdist+1));
+			}
+
+		}
+
+	}
+
+	private static boolean ischeckRotate(int cy, int cx, int nstate) {
+		if(visit[cy][cx][nstate]) return false;
+		for(int i=0;i<8;i++) {
+			int ny = cy+dy[i];
+			int nx = cx+dx[i];
+			if(ny<0||nx<0||ny>=N||nx>=N) return false;
+			if(board[ny][nx]=='1') return false;
+		}
+		
+		return true;
+	}
+
+	private static boolean ischeck(int ny, int nx, int cstate) {
+		
+		if(ny<0||nx<0||ny>=N||nx>=N) return false;
+		if(visit[ny][nx][cstate]) return false;
+		
+		if(cstate==0) {
+			for(int i=0;i<3;i++) {
+				int tx = nx+lcr[i];
+				if(tx<0||tx>=N) return false;
+				if(board[ny][tx]=='1') return false;
+			}
+		}
+		
+		else {
+			for(int i=0;i<3;i++) {
+				int ty = ny+lcr[i];
+				if(ty<0||ty>=N) return false;
+				if(board[ty][nx]=='1') return false;
+			}
+		}
```

Comment:
`cstate` 기준으로 값을 할당하면 중복 코드를 줄일 수 있을 것 같아요
```suggestion
	for(int i=0;i<3;i++) {
		int ty = ny;
		int tx = nx;
		
		if(cstate == 0)		tx += lcr[i];
		else			ty += lcr[i];
		 
		if(tx < 0 || tx >= N || ty < 0 || ty >= N) return false;
		if(board[ty][tx]=='1') return false;
	}

```